--- a/drivers/gpio/gpiolib-of.c
+++ b/drivers/gpio/gpiolib-of.c
@@ -26,7 +26,7 @@ static int of_gpiochip_match_node_and_xlate(struct gpio_chip *chip, void *data)
 {
 	struct of_phandle_args *gpiospec = data;
 
-	return chip->gpiodev->dev.of_node == gpiospec->np &&
+	return (chip->of_node == gpiospec->np || chip->gpiodev->dev.of_node == gpiospec->np) &&
 				chip->of_xlate(chip, gpiospec, NULL) >= 0;
 }
 
